internal/graph/object_providers_test.go

```diff
@@ -454,98 +453,111 @@ func TestRecursiveUsersetObjectProvider(t *testing.T) {
+		iterChan3 := make(chan *iterator.Msg, 1)
+		iterChan3 <- &iterator.Msg{Iter: storage.NewStaticIterator[string]([]string{"7"})}
+		close(iterChan3)
+		chans = append(chans, iterChan3)
+
+		outChan := make(chan usersetMessage, len(chans))
+		ctx := context.Background()
+		go iteratorsToUserset(ctx, chans, outChan)
+		count := 0
+		for msg := range outChan {
+			if msg.err != nil {
+				if errors.Is(msg.err, mocks.ErrSimulatedError) {
+					continue
 				}
+				require.Fail(t, "unexpected error", msg.err)
 			}
-			close(inputMessages)
-			wg := sync.WaitGroup{}
-			wg.Add(1)
-			var output []usersetMessage
-			outChan := make(chan usersetMessage)
-			go func() {
-				defer wg.Done()
-				for msg := range outChan {
-					output = append(output, msg)
-				}
-			}()
-			err := iteratorToUserset(inputMessages, outChan)(tt.ctx)
-			require.Equal(t, tt.expectedErr, err)
-
-			wg.Wait()
-			require.Equal(t, tt.expectedResult, output)
+			id, err := strconv.Atoi(msg.userset)
+			require.NoError(t, err)
+			require.Positive(t, id)
+			require.LessOrEqual(t, id, 7)
+			count++
+		}
+		require.Equal(t, 7, count)
+	})
+	t.Run("cancellation", func(t *testing.T) {
+		t.Cleanup(func() {
+			// this is the expected goroutine
+			goleak.VerifyNone(t, goleak.IgnoreTopFunction("github.com/openfga/openfga/internal/iterator.Drain.func1"))
+		})
+		chans := make([]chan *iterator.Msg, 0, 5)
+		for i := 1; i <= 5; i++ {
+			iterChan := make(chan *iterator.Msg, 1)
+			iterChan <- &iterator.Msg{Iter: storage.NewStaticIterator[string]([]string{strconv.Itoa(i)})}
+			// close(iterChan) -> by not closing, ctx.Done() is the exit clause
+			chans = append(chans, iterChan)
+		}
+		outChan := make(chan usersetMessage, len(chans))
+		ctx := context.Background()
+		ctx, cancel := context.WithCancel(ctx)
+		cancel()
+		go iteratorsToUserset(ctx, chans, outChan)
+
+		count := 0
+		for msg := range outChan {
+			id, err := strconv.Atoi(msg.userset)
+			require.NoError(t, err)
+			require.Positive(t, id)
+			require.LessOrEqual(t, id, 5)
+			count++
+		}
+		require.LessOrEqual(t, count, 5)
+	})
+	t.Run("handles_errors", func(t *testing.T) {
+		t.Cleanup(func() {
+			// this is the expected goroutine due to "iterator error"
+			goleak.VerifyNone(t, goleak.IgnoreTopFunction("github.com/openfga/openfga/internal/iterator.Drain.func1"))
 		})
-	}
+		chans := make([]chan *iterator.Msg, 0, 2)
+		iterChan1 := make(chan *iterator.Msg, 1)
+		iterChan1 <- &iterator.Msg{Err: errors.New("iterator error")}
+		close(iterChan1)
+		chans = append(chans, iterChan1)
+		iterChan2 := make(chan *iterator.Msg, 1)
+		iterChan2 <- &iterator.Msg{Iter: mocks.NewErrorIterator[string]([]string{"1"})}
+		close(iterChan2)
+		chans = append(chans, iterChan2)
+
+		outChan := make(chan usersetMessage, len(chans))
+		ctx := context.Background()
+		go iteratorsToUserset(ctx, chans, outChan)
+
+		count := 0
+		for msg := range outChan {
+			if msg.err != nil {
+				count++
+			}
+		}
+		require.Equal(t, 2, count)
+	})
```

Comment:
should we verify that we see the message content (i.e., there is at least 1 error)?

Reply:
We are actually counting the errors themselves 🤔